### PR TITLE
guard-sevice

### PR DESCRIPTION
**Pull request by davidhadas**
This PR finalizes the guard-service

It also includes examples of a Guardian CRD and a Guardian ConfigMap as a reference

### CODE REVIEW

File: cmd/guard-service/main.go [hunk 1]
```diff
@@ -45,12 +49,74 @@ type learner struct {
 	pileLearnTicker *utils.Ticker
 }
 
+// Common method used for parsing ns, sid, cmFlag from all requests
+func (l *learner) baseHandler(query url.Values) (ns string, sid string, cmFlag bool, record *serviceRecord) {
```
Comment (maximilien) [inline cmd/guard-service/main.go:53]: Why not return an error also in this function and that way deal with it from the caller's side?

File: cmd/guard-service/main.go [hunk 2]
```diff
@@ -45,12 +49,74 @@ type learner struct {
...
+		log.Infof("baseHandler illegal sid")
+		sid = ""
+		return
+	}
+
+	record = l.services.get(ns, sid, cmFlag)
+	return
+}
+
 func (l *learner) fetchConfig(w http.ResponseWriter, req *http.Request) {
```
Comment (maximilien) [inline cmd/guard-service/main.go:76]: Same as previous function. return error and deal with it in the caller code. This allows more flexibility for retries etc.

File: cmd/guard-service/services.go [hunk 3]
```diff
@@ -49,10 +49,16 @@ func serviceKey(ns string, sid string, cmFlag bool) string {
 	return service
 }
 
-func newServices() *services {
+func _newServices() *services {
```
Comment (maximilien) [inline cmd/guard-service/services.go:52]: Hmm, wondering why using _ in the name... I know it's private but Golang has strong formatting rules and people expect you to follow them. At a minimum add a comment to explain your reasoning.

File: pkg/guard-utils/santize.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+limitations under the License.
+*/
+
+package guardutils
+
+// contain at most 63 characters
+// contain only lowercase alphanumeric characters or '-'
+// start with an alphanumeric character
+// end with an alphanumeric character
+func Sanitize(in string) string {
```
Comment (maximilien) [inline pkg/guard-utils/santize.go:23]: I'd separate the different sub-sections into helper functions (easy) and Sanatize just calls the others. Allows for easy testing and easier modification if Sanatize needs to be enhanced...

**Review by maximilien — COMMENTED** [review]
Left a few comments. Thanks for constant progress on this and for keeping PRs small :)

File: pkg/guard-utils/santize.go (hunk 4 above)
Comment (davidhadas) [reply pkg/guard-utils/santize.go:23]: let's let the natural course of things define what other sanitization is needed and build helpers as needed and where needed.

**Comment by davidhadas** [comment]
/hold
Merge #58 first to make this PR simpler

**Review by maximilien — APPROVED** [review]
/approve
/lgtm

**Pull request merged**